JavaScript annotation doesn't work with context:// URLs in npm mode

I have the `@JavaScript("context://js/test.js")` attached to my route, in hopes to load JavaScript files as plain JS (not as a module) from `src/main/webapp/js/test.js`.

However, Vaadin tries to look up the javascript file from the `node_modules` and fails.


### Minimal reproducible example

1. The Skeleton Starter project: Add `@JavaScript("context://js/test.js")` annotation to the `MainView` class.
2. Run the Skeleton Starter.

### Expected behavior

The produced HTML page should load a script from `src/main/webapp/js/test.js`.

### Actual behavior

The webpack compilation fails with

```
 Failed to find the following imports in the `node_modules` tree:
      - context://js/test.js
  If the build fails, check that npm packages are installed.
```

### Versions:
    - Vaadin / Flow version: 4.1.27 / 2.1.9
    - Java version: 8
    - OS version:
    - Browser version (if applicable):
    - Application Server (if applicable):
    - IDE (if applicable): Intellij

Full console log:
```
...
2020-05-11 11:46:50.860 [RMI TCP Connection(2)-127.0.0.1] INFO com.vaadin.flow.server.startup.DevModeInitializer - Starting dev-mode updaters in /home/mavi/work/my/vok/karibu10-helloworld-application folder.
2020-05-11 11:46:50.921 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - Visited 63 classes. Took 42 ms.
2020-05-11 11:46:51.024 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - Skipping `npm install`.
2020-05-11 11:46:51.024 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - Copying frontend resources from jar files ...
2020-05-11 11:46:51.049 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - Visited 11 resources. Took 25 ms.
2020-05-11 11:46:51.082 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - 

  Failed to find the following imports in the `node_modules` tree:
      - context://js/test.js
  If the build fails, check that npm packages are installed.

  To fix the build remove `package-lock.json` and `node_modules` directory to reset modules.
  In addition you may run `npm install` to fix `node_modules` tree structure.


2020-05-11 11:46:51.085 [RMI TCP Connection(2)-127.0.0.1] INFO dev-updater - No js modules to update '/home/mavi/work/my/vok/karibu10-helloworld-application/target/frontend/generated-flow-imports.js' file
2020-05-11 11:46:52.203 [RMI TCP Connection(2)-127.0.0.1] INFO dev-webpack - Starting webpack-dev-server, port: 40673 dir: /home/mavi/work/my/vok/karibu10-helloworld-application
2020-05-11 11:46:52.205 [RMI TCP Connection(2)-127.0.0.1] INFO dev-webpack - Running webpack to compile frontend resources. This may take a moment, please stand by...
2020-05-11 11:46:57.810 [webpack] ERROR dev-webpack - ERROR in ../target/frontend/generated-flow-imports.js
2020-05-11 11:46:57.810 [webpack] ERROR dev-webpack - Module not found: Error: Can't resolve 'context://js/test.js' in '/home/mavi/work/my/vok/karibu10-helloworld-application/target/frontend'
2020-05-11 11:46:57.810 [webpack] ERROR dev-webpack -  @ ../target/frontend/generated-flow-imports.js 85:0-30
2020-05-11 11:46:57.810 [webpack] ERROR dev-webpack - ERROR in ../target/frontend/generated-flow-imports.js
2020-05-11 11:46:57.810 [webpack] ERROR dev-webpack - Module not found: Error: Can't resolve 'context://js/test.js' in '/home/mavi/work/my/vok/karibu10-helloworld-application/target/frontend'
2020-05-11 11:46:57.811 [webpack] ERROR dev-webpack -  @ ../target/frontend/generated-flow-imports.js 85:0-30
2020-05-11 11:46:57.811 [webpack] ERROR dev-webpack - ℹ ｢wdm｣: Failed to compile.
2020-05-11 11:46:57.811 [webpack] ERROR dev-webpack - 
------------------ Frontend compilation failed. -----------------
2020-05-11 11:46:57.811 [RMI TCP Connection(2)-127.0.0.1] INFO dev-webpack - Webpack startup and compilation completed in 5608ms
2020-05-11 11:46:57.827 [RMI TCP Connection(2)-127.0.0.1] INFO com.vaadin.flow.server.startup.ServletDeployer - Automatically deploying Vaadin servlet with name com.vaadin.flow.server.startup.ServletDeployer to /*
[2020-05-11 11:46:58,052] Artifact Gradle : karibu10-helloworld-application.war (exploded): Artifact is deployed successfully
[2020-05-11 11:46:58,052] Artifact Gradle : karibu10-helloworld-application.war (exploded): Deploy took 9,033 milliseconds
2020-05-11 11:46:58.637 [http-nio-8080-exec-1] WARN com.vaadin.flow.server.DefaultDeploymentConfiguration - 
====================================================================
Vaadin is running in DEBUG MODE.
In order to run your application in production mode and disable debug features, you should enable it by setting the servlet init parameter productionMode to true.
See https://vaadin.com/docs/v14/flow/production/tutorial-production-mode-basic.html for more information about the production mode.
====================================================================
/home/mavi/local/apache-tomcat-9.0.31/bin/catalina.sh stop
Disconnected from the target VM, address: '127.0.0.1:37955', transport: 'socket'
```

--- a/flow-server/src/main/java/com/vaadin/flow/component/dependency/JavaScript.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/dependency/JavaScript.java
@@ -67,9 +67,18 @@ public @interface JavaScript {
      * in the browser.
      * <p>
      * Relative URLs are interpreted as relative to the configured
-     * {@code frontend} directory location. You can prefix the URL with
-     * {@code context://} to make it relative to the context path or use an
-     * absolute URL to refer to files outside the frontend directory.
+     * {@code frontend} directory location.
+     * <p>
+     * In NPM mode this URL identifies a file which will be bundled, so the file
+     * should be available to be able to bundle it. External URLs (e.g. absolute
+     * URL which has a schema like "http://") are not bundled but included into
+     * the page as standalone scripts (the same way as in compatibility mode,
+     * see below). It's applicable also for "context://" and "base://" schemas.
+     * <p>
+     * In compatibility mode the URL represents a standalone script which will
+     * be added to the page. You can prefix the URL with {@code context://} to
+     * make it relative to the context path or use an absolute URL to refer to
+     * files outside the frontend directory.
      *
      * @return a JavaScript file URL
      */

--- a/flow-server/src/main/java/com/vaadin/flow/internal/UrlUtil.java
+++ b/flow-server/src/main/java/com/vaadin/flow/internal/UrlUtil.java
@@ -41,9 +41,6 @@ public class UrlUtil {
         if (url.startsWith("//")) {
             return true;
         }
-        return url.contains("://") && !(url
-                .startsWith(ApplicationConstants.CONTEXT_PROTOCOL_PREFIX)
-                || url.startsWith(ApplicationConstants.FRONTEND_PROTOCOL_PREFIX)
-                || url.startsWith(ApplicationConstants.BASE_PROTOCOL_PREFIX));
+        return url.contains("://") && !url.startsWith(ApplicationConstants.FRONTEND_PROTOCOL_PREFIX);
     }
 }

--- a/flow-server/src/test/java/com/vaadin/flow/internal/UrlUtilTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/internal/UrlUtilTest.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.internal;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class UrlUtilTest {
+
+    @Test
+    public void isExternal_URLStartsWithTwoSlashes_returnsTrue() {
+        Assert.assertTrue(UrlUtil.isExternal("//foo"));
+    }
+
+    @Test
+    public void isExternal_URLContainsAnySchemaAsPrefix_returnsTrue() {
+        Assert.assertTrue(UrlUtil.isExternal("http://foo"));
+        Assert.assertTrue(UrlUtil.isExternal("https://foo"));
+        Assert.assertTrue(UrlUtil.isExternal("context://foo"));
+        Assert.assertTrue(UrlUtil.isExternal("base://foo"));
+    }
+
+    @Test
+    public void isExternal_URLDoesnotContainSchema_returnsFalse() {
+        Assert.assertFalse(UrlUtil.isExternal("foo"));
+    }
+}

--- a/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/PlainScriptViaJavaScriptView.java
+++ b/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/PlainScriptViaJavaScriptView.java
@@ -1,0 +1,26 @@
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.uitest.ui;
+
+import com.vaadin.flow.component.dependency.JavaScript;
+import com.vaadin.flow.component.html.Div;
+import com.vaadin.flow.router.Route;
+
+@Route(value = "com.vaadin.flow.uitest.ui.PlainScriptViaJavaScriptView")
+@JavaScript("context://plain-script.js")
+public class PlainScriptViaJavaScriptView extends Div {
+
+}

--- a/flow-tests/test-root-context/src/main/webapp/plain-script.js
+++ b/flow-tests/test-root-context/src/main/webapp/plain-script.js
@@ -1,0 +1,4 @@
+var div = document.createElement("div");
+div.id="added-from-src-script";
+div.textContent="Hello from src script";
+document.body.insertBefore(div, document.body.firstChild);

--- a/flow-tests/test-root-context/src/test/java/com/vaadin/flow/uitest/ui/PlainScriptViaJavaScriptIT.java
+++ b/flow-tests/test-root-context/src/test/java/com/vaadin/flow/uitest/ui/PlainScriptViaJavaScriptIT.java
@@ -1,0 +1,32 @@
+/*
+ * Copyright 2000-2020 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.uitest.ui;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.openqa.selenium.By;
+
+import com.vaadin.flow.testutil.ChromeBrowserTest;
+
+public class PlainScriptViaJavaScriptIT extends ChromeBrowserTest {
+
+    @Test
+    public void contextSchemaWorksinJavaScript() {
+        open();
+
+        Assert.assertTrue(isElementPresent(By.id("added-from-src-script")));
+    }
+}

